src/bech32.c

```diff
@@ -165,36 +179,47 @@ static int convert_bits(uint8_t *out, size_t *outlen, int outbits, const uint8_t
     return 1;
 }
 
-static int segwit_addr_encode(char *output, const char *hrp, int witver, const uint8_t *witprog, size_t witprog_len) {
+
+int segwit_addr_encode(char *output, const char *hrp, int witver, const uint8_t *witprog, size_t witprog_len) {
     uint8_t data[65];
     size_t datalen = 0;
-    if (witver < 0 || witver > 16) goto fail;
-    if (witver == 0 && witprog_len != 20 && witprog_len != 32) goto fail;
-    if (witprog_len < 2 || witprog_len > 40) goto fail;
-    data[0] = witver;
-    convert_bits(data + 1, &datalen, 5, witprog, witprog_len, 8, 1);
-    ++datalen;
-    return bech32_encode(output, hrp, data, datalen, 90);
-fail:
+    bech32_encoding enc = BECH32_ENCODING_BECH32;
+    do {
+        if (witver > 16) break;
+        if (witver == 0 && witprog_len != 20 && witprog_len != 32) break;
+        if (witprog_len < 2 || witprog_len > 40) break;
+        if (witver > 0) enc = BECH32_ENCODING_BECH32M;
+        data[0] = witver;
+        convert_bits(data + 1, &datalen, 5, witprog, witprog_len, 8, 1);
+        ++datalen;
+        return bech32_encode(output, hrp, data, datalen, enc);
+    } while(0);
+
     wally_clear_2(data, sizeof(data), (void *)witprog, witprog_len);
     return 0;
 }
 
-static int segwit_addr_decode(int *witver, uint8_t *witdata, size_t *witdata_len, const char *hrp, const char *addr) {
+
+int segwit_addr_decode(int* witver, uint8_t* witdata, size_t* witdata_len, const char* hrp, const char* addr) {
     uint8_t data[84];
     char hrp_actual[84];
     size_t data_len;
-    if (!bech32_decode(hrp_actual, data, &data_len, addr, 90)) goto fail;
-    if (data_len == 0 || data_len > 65) goto fail;
-    if (strncmp(hrp, hrp_actual, 84) != 0) goto fail;
-    if (data[0] > 16) goto fail;
-    *witdata_len = 0;
-    if (!convert_bits(witdata, witdata_len, 8, data + 1, data_len - 1, 5, 0)) goto fail;
-    if (*witdata_len < 2 || *witdata_len > 40) goto fail;
-    if (data[0] == 0 && *witdata_len != 20 && *witdata_len != 32) goto fail;
-    *witver = data[0];
-    return 1;
-fail:
+    do {
```

Comment:
goto is not evil, please do not reformulate control structures without need, it complicates the review process.